src/box/replication.cc

```diff
@@ -314,8 +313,14 @@ replica_delete(struct replica *replica)
 	assert(replica_is_orphan(replica));
 	if (replica->relay != NULL)
 		relay_delete(replica->relay);
-	if (replica->gc != NULL)
-		gc_consumer_unregister(replica->gc);
+	/*
+	 * Drop consumer only if another replica does not own it.
+	 * Another replica can own it, for example, when we created a new
+	 * replica object on `replicaset_update`, and now we drop it because
+	 * we can reuse the old one.
+	 */
+	if (replica_by_uuid(&replica->uuid) == NULL)
+		gc_consumer_unregister(&replica->uuid);
 	TRASH(replica);
```

Comment:
That is a bit sad that the consumer ownership is still here, but now it is implicit. Could you keep `gc_consumer` pointer being explicitly stored in the replica like before? I mean, nothing changed really in this commit besides that now we don't see who owns what, and have to do hacks like this. You said *"their lifetime became independent"*, but it didn't really.

Having the pointers didn't seem to block the anon consumers anyhow.

I would really appreciate if we would find a way to keep the "ownership" concept. Too easy to break things otherwise.

Also, if we allow to have multiple consumers for one UUID at once, we should mark one of them as the main one, and others would be orphaned and their change wouldn't affect the persisted state. Otherwise there needs to be an explicit protection from having more than one consumer object for the same UUID (assertions for impossible cases, runtime checks for client errors - currently, for example, I see that all the rb_tree_insert calls are not checked whether they failed or not).

Reply:
Completely refactored the patch, so the comment is outdated.